corehq/apps/reports/v2/endpoints/case_properties.py

```diff
@@ -37,16 +53,66 @@ def case_properties(self):
         ]
         return case_properties + special_properties
 
+    @staticmethod
+    def _fmt_select2(prop):
+        prop['text'] = prop['name']
+        prop['id'] = prop['name']
+        prop['labelStyle'] = ('label-default' if prop['case_type']
+                              else 'label-primary')
+        prop['labelText'] = prop['case_type'] or prop['meta_type']
+        return prop
+
+
     @property
-    def filtered_case_properties(self):
+    @memoized
+    def matching_case_properties(self):
+        exact = []
+        starts = []
+        ends = []
+        other = []
+
+        for prop in self.case_properties:
+            if prop['name'] in self.excluded_slugs:
+                continue
+
+            if not self.search:
+                exact.append(self._fmt_select2(prop))
+
+            elif prop['name'] == self.search:
+                exact.append(self._fmt_select2(prop))
+
+            elif prop['name'].startswith(self.search):
+                starts.append(self._fmt_select2(prop))
 
-        def _filter_property(prop):
-            # todo future filters based on case_types, etc.
-            return prop['name'] not in self.excluded_slugs
+            elif prop['name'].endswith(self.search):
+                ends.append(self._fmt_select2(prop))
 
-        return [prop for prop in self.case_properties if _filter_property(prop)]
+            elif self.search in prop['name']:
+                other.append(self._fmt_select2(prop))
+
+        if len(exact) == 0 and self.search:
+            exact.append({
+                'text': self.search,
+                'id': self.search,
+                'labelStyle': 'label-info',
+                'labelText': '({})'.format(_("custom")),
+            })
+
+        return exact + starts + ends + other
 
     def get_response(self):
+        #  todo more intelligent caching of results
```

Comment:
Did you think about just passing everything to the client side every time and filtering with select2? The number of items is never going to be huge. That way it doesn't have to do a round trip to the server every time you type something, which seems slow (even in your example screenshot, there is a lag).

Reply:
i did actually think about that, but I felt that it would make it much more difficult to generalize the report. also what happens when a ton of data has to be stored on the client side? that really slows things down too...

Reply:
I'm not sure the volume of data could ever be large enough to slow this down more than making a server round trip on a slow connection for every few keypresses. For a regular SaaS project people probably have in the 100s of different case property names. Filtering a list of that order of magnitude should be "fast". Even in the extreme case where a domain has 1000s of case properties, sending a list with 5000 strings in it once can hardly be called "a ton of data".